File: src/use-cases/party/CreateParty.js
/**
 * CreateParty Use Case
 * Creates a new party with specified settings
 */

const Party = require('../../domain/entities/Party');
const PartySettings = require('../../domain/value-objects/PartySettings');
const logger = require('../../../logger');

class CreateParty {
    /**
     * @param {IPartyRepository} partyRepository - Party repository
     * @param {IUserRepository} userRepository - User repository
     * @param {JoinParty} joinPartyUseCase - JoinParty use case for adding bots
     */
    constructor(partyRepository, userRepository, joinPartyUseCase) {
        this.partyRepository = partyRepository;
        this.userRepository = userRepository;
        this.joinPartyUseCase = joinPartyUseCase;
    }

    /**
     * Execute the use case
     * @param {Object} request - Party creation request
     * @param {string} request.ownerId - User ID of party owner
     * @param {string} request.name - Party name
     * @param {string} request.visibility - 'public' or 'private'
     * @param {Object} request.settings - Party settings
     * @param {number} request.settings.playerCount - Number of players (3-8)
     * @param {number} request.settings.handSize - Cards per hand (5-7)
     * @param {boolean} [request.settings.allowSpectators] - Allow spectators
     * @param {number} [request.settings.roundTimeLimit] - Time limit per round in seconds
     * @param {string[]} [request.botIds] - Array of bot user IDs to auto-join
     * @returns {Promise<Object>} Creation result with party
     */
    async execute({ ownerId, name, visibility, settings, botIds = [] }) {
        try {
            // Validate input
            if (!ownerId || typeof ownerId !== 'string') {
                throw new Error('Owner ID is required');
            }

            if (!name || typeof name !== 'string') {
                throw new Error('Party name is required');
            }

            const trimmedName = name.trim();
            if (trimmedName.length < 3) {
                throw new Error('Party name must be at least 3 characters long');
            }

            if (trimmedName.length > 50) {
                throw new Error('Party name must not exceed 50 characters');
            }

            if (!visibility || !['public', 'private'].includes(visibility)) {
                throw new Error('Visibility must be either "public" or "private"');
            }

            if (!settings || typeof settings !== 'object') {
                throw new Error('Settings are required');
            }

            // Verify owner exists
            const owner = await this.userRepository.findById(ownerId);
            if (!owner) {
                throw new Error('Owner not found');
            }

            // Create party settings
            const partySettings = new PartySettings({
                playerCount: settings.playerCount,
                handSize: settings.handSize,
                allowSpectators: settings.allowSpectators || false,
                roundTimeLimit: settings.roundTimeLimit || 0
            });

            // Validate botIds if provided
            if (botIds && botIds.length > 0) {
                // Check botIds is array
                if (!Array.isArray(botIds)) {
                    throw new Error('botIds must be an array');
                }

                // Check for duplicates
                const uniqueBotIds = new Set(botIds);
                if (uniqueBotIds.size !== botIds.length) {
                    throw new Error('Duplicate bot IDs detected');
                }

                // Check total players doesn't exceed player count (owner + bots)
                if (botIds.length + 1 > partySettings.playerCount) {
                    throw new Error(`Total players (${botIds.length + 1}) exceeds party player count (${partySettings.playerCount})`);
                }

                // Verify all bot IDs exist and are actual bots
                for (const botId of botIds) {
                    const bot = await this.userRepository.findById(botId);
                    if (!bot) {
                        throw new Error(`Bot with ID ${botId} not found`);
                    }
                    if (!bot.isBot()) {
                        throw new Error(`User ${botId} is not a bot`);
                    }
                }
            }

            // Create party
            const party = Party.create(trimmedName, ownerId, visibility, partySettings);

            // Save party
            const savedParty = await this.partyRepository.save(party);

            logger.info('Party created successfully', {
                partyId: savedParty.id,
                name: savedParty.name,
                ownerId: savedParty.ownerId,
                visibility: savedParty.visibility,
                botCount: botIds.length
            });

            // Auto-join bots if provided
            if (botIds && botIds.length > 0 && this.joinPartyUseCase) {
                logger.info('Auto-joining bots to party', {
                    partyId: savedParty.id,
                    botIds
                });

                const joinedBots = [];
                for (const botId of botIds) {
                    try {
                        await this.joinPartyUseCase.execute({
                            userId: botId,
                            partyId: savedParty.id
                        });
                        joinedBots.push(botId);
                        logger.info('Bot joined party', { botId, partyId: savedParty.id });
                    } catch (error) {
                        logger.error('Failed to auto-join bot', {
                            botId,
                            partyId: savedParty.id,
                            error: error.message
                        });
                        // Continue with other bots even if one fails
                    }
                }

                // Reload party to get updated player list
                const updatedParty = await this.partyRepository.findById(savedParty.id);

                // Ensure party stays in 'waiting' status even if all slots filled
                // (User requirement: always wait for manual start)
                if (updatedParty.status === 'playing') {
                    updatedParty.status = 'waiting';
                    await this.partyRepository.save(updatedParty);
                }

                logger.info('Bots auto-joined successfully', {
                    partyId: savedParty.id,
                    requestedCount: botIds.length,
                    joinedCount: joinedBots.length
                });

                return {
                    success: true,
                    party: updatedParty.toPublicObject(),
                    botsJoined: joinedBots.length
                };
            }

            return {
                success: true,
                party: savedParty.toPublicObject()
            };
        } catch (error) {
            logger.error('Party creation error', {
                ownerId,
                name,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = CreateParty;

File: zapzap-rust/src/application/party/create_party.rs
use std::sync::Arc;

use uuid::Uuid;

use crate::domain::entities::{generate_invite_code, Party, PartyVisibility};
use crate::domain::repositories::{PartyRepository, RepositoryError, UserRepository};
use crate::domain::value_objects::PartySettings;

/// Party name bounds once trimmed, in UTF-16 code units as JavaScript's `length`
/// (Node: CreateParty.js), so "🎲🎲" is 4 long
pub const MIN_NAME_LENGTH: usize = 3;
pub const MAX_NAME_LENGTH: usize = 50;

/// Create party input
pub struct CreatePartyInput {
    pub owner_id: String,
    pub name: String,
    pub visibility: String,
    pub settings: PartySettings,
    pub bot_ids: Vec<String>,
}

/// Create party output
pub struct CreatePartyOutput {
    pub party: Party,
    pub bots_joined: usize,
}

/// Create party use case
pub struct CreateParty<U: UserRepository, P: PartyRepository> {
    user_repo: Arc<U>,
    party_repo: Arc<P>,
}

impl<U: UserRepository, P: PartyRepository> CreateParty<U, P> {
    pub fn new(user_repo: Arc<U>, party_repo: Arc<P>) -> Self {
        Self {
            user_repo,
            party_repo,
        }
    }

    pub async fn execute(
        &self,
        input: CreatePartyInput,
    ) -> Result<CreatePartyOutput, CreatePartyError> {
        // Validate user exists
        let user = self
            .user_repo
            .find_by_id(&input.owner_id)
            .await?
            .ok_or(CreatePartyError::UserNotFound)?;

        if user.is_bot() {
            return Err(CreatePartyError::Validation(
                "Bots cannot create parties".into(),
            ));
        }

        // Validate name, trimmed, as Node's CreateParty.js
        let name = input.name.trim().to_string();
        let name_length = name.encode_utf16().count();
        if name_length < MIN_NAME_LENGTH {
            return Err(CreatePartyError::Validation(format!(
                "Party name must be at least {MIN_NAME_LENGTH} characters long"
            )));
        }
        if name_length > MAX_NAME_LENGTH {
            return Err(CreatePartyError::Validation(format!(
                "Party name must not exceed {MAX_NAME_LENGTH} characters"
            )));
        }

        input
            .settings
            .validate()
            .map_err(CreatePartyError::Validation)?;

        // Node: CreateParty.js
        let unique_bot_ids: std::collections::HashSet<&String> = input.bot_ids.iter().collect();
        if unique_bot_ids.len() != input.bot_ids.len() {
            return Err(CreatePartyError::Validation(
                "Duplicate bot IDs detected".into(),
            ));
        }

        // The owner and the bots must fit in the seats (Node: CreateParty.js)
        let total_players = input.bot_ids.len() + 1;
        if total_players > input.settings.player_count as usize {
            return Err(CreatePartyError::Validation(format!(
                "Total players ({total_players}) exceeds party player count ({})",
                input.settings.player_count
            )));
        }

        // Parse visibility
        let visibility =
            PartyVisibility::from_str(&input.visibility).unwrap_or(PartyVisibility::Public);

        // Create party
        let party_id = Uuid::new_v4().to_string();
        let invite_code = generate_invite_code();

        let party = Party::new(
            party_id.clone(),
            name,
            input.owner_id.clone(),
            invite_code,
            visibility,
            input.settings,
        );

        // Save party
        self.party_repo.save(&party).await?;

        // Add owner as first player
        self.party_repo
            .add_party_player(&party_id, &input.owner_id, 0)
            .await?;

        let mut bots_joined = 0;

        // Add bots if specified: contiguous seats after the owner, whatever ids are skipped
        for bot_id in &input.bot_ids {
            // Verify bot exists
            if let Some(bot) = self.user_repo.find_by_id(bot_id).await? {
                if bot.is_bot() {
                    self.party_repo
                        .add_party_player(&party_id, bot_id, (bots_joined + 1) as u8)
                        .await?;
                    bots_joined += 1;
                }
            }
        }

        Ok(CreatePartyOutput { party, bots_joined })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum CreatePartyError {
    #[error("Validation error: {0}")]
    Validation(String),
    #[error("User not found")]
    UserNotFound,
    #[error("Repository error: {0}")]
    Repository(#[from] RepositoryError),
}
